Subject: [PATCH] fix: 提高一些绘制效率
---
Index: app/src/main/protobuf/OnlineRequestProtocol.proto
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/app/src/main/protobuf/OnlineRequestProtocol.proto b/app/src/main/protobuf/OnlineRequestProtocol.proto
--- a/app/src/main/protobuf/OnlineRequestProtocol.proto	(revision 7ac040dd7682e2677690e16035d771318d8ed6fc)
+++ b/app/src/main/protobuf/OnlineRequestProtocol.proto	(date 1694598513873)
@@ -185,6 +185,8 @@
   string message = 2;
   // 消息文字颜色
   uint32 color = 3;
+  // 消息文字颜色
+  uint32 type = 4;
 }
 
 // 改变房间信息
Index: app/src/main/java/ly/pp/justpiano3/activity/OLPlayRoomActivity.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/app/src/main/java/ly/pp/justpiano3/activity/OLPlayRoomActivity.java b/app/src/main/java/ly/pp/justpiano3/activity/OLPlayRoomActivity.java
--- a/app/src/main/java/ly/pp/justpiano3/activity/OLPlayRoomActivity.java	(revision 7ac040dd7682e2677690e16035d771318d8ed6fc)
+++ b/app/src/main/java/ly/pp/justpiano3/activity/OLPlayRoomActivity.java	(date 1694598353936)
@@ -4,10 +4,7 @@
 import android.os.*;
 import android.text.Selection;
 import android.text.Spannable;
-import android.view.Gravity;
-import android.view.LayoutInflater;
-import android.view.View;
-import android.view.WindowManager;
+import android.view.*;
 import android.widget.*;
 import com.google.protobuf.MessageLite;
 import ly.pp.justpiano3.JPApplication;
@@ -37,7 +34,7 @@
 /**
  * 房间
  */
-public class OLPlayRoomActivity extends OLBaseActivity implements Handler.Callback, View.OnClickListener {
+public class OLPlayRoomActivity extends OLBaseActivity implements Handler.Callback, View.OnClickListener, View.OnLongClickListener {
     public int lv;
     public int cl;
     public Handler handler;
@@ -223,7 +220,7 @@
         jpdialog.showDialog();
     }
 
-    protected void sendMailClick() {
+    protected void sendMailClick(boolean isBroadcast) {
         String str;
         OnlineRoomChatDTO.Builder builder2 = OnlineRoomChatDTO.newBuilder();
         str = String.valueOf(sendText.getText());
@@ -538,6 +535,7 @@
         msgListView.setCacheColorHint(0);
         Button olSendButton = findViewById(R.id.ol_send_b);
         olSendButton.setOnClickListener(this);
+        olSendButton.setOnLongClickListener(this);
         timeTextView = findViewById(R.id.time_text);
         findViewById(R.id.pre_button).setOnClickListener(this);
         findViewById(R.id.next_button).setOnClickListener(this);
@@ -638,7 +636,7 @@
                 nextFriendPageClick();
                 return;
             case R.id.ol_send_b:
-                sendMailClick();
+                sendMailClick(false);
                 return;
             case R.id.ol_express_b:
                 expressWindow.showAtLocation(expressImageView, Gravity.CENTER, 0, 0);
@@ -682,4 +680,23 @@
             default:
         }
     }
+
+    @Override
+    public boolean onLongClick(View v) {
+        int vid = v.getId();
+        if (vid == R.id.ol_send_b) {
+            new JPDialog(this)
+                    .setTitle("发送'全服广播'")
+                    .setMessage("您是否希望使用'全服广播'进行发送？\n如果您没有，则会自动为您购买并发送。\n内容为：" + sendText.getText())
+                    .setFirstButton("我希望", (dialog, i) -> {
+                        sendMailClick(true);
+                    })
+                    .setSecondButton("再想想", new DialogDismissClick())
+                    .showDialog();
+            return true;
+        }
+
+
+        return false;
+    }
 }
